File: myAlgorithm/001冒泡排序.js
function bubbleSort(array){
  if(typeof array != "object"){
    return array;
  }
  if(array.length  <= 1){
    return array;
  }
  for(var i = 0; i < array.length - 1; i++){
    for(var j = 0; j < array.length - 1 - i; j++){
      if(array[j + 1] < array[j]){
        var temp = array[j + 1];
        array[j + 1] = array[j];
        array[j] = temp;
      }
    }
  }
  return array;
}

function bubbleSort(array){
  if(typeof array != "object"){
    return array;
  }
  if(array.length  <= 1){
    return array;
  }
  for(var i = 0; i < array.length - 1; i++){
    var isSorted = true;
    for(var j = 0; j < array.length - 1 - i; j++){
      if(array[j + 1] < array[j]){
        var temp = array[j + 1];
        array[j + 1] = array[j];
        array[j] = temp;
        isSorted = false;
      }
    }
    if(isSorted){
      break;
    }
  }
  return array;
}

function bubbleSort(array){
  if(typeof array != "object"){
    return array;
  }
  if(array.length  <= 1){
    return array;
  }
  var lastExchangeIndex = 0;
  var sortedBorder = array.length - 1;
  for(var i = 0; i < array.length - 1; i++){
    var isSorted = true;
    for(var j = 0; j < sortedBorder; j++){
      if(array[j + 1] < array[j]){
        var temp = array[j + 1];
        array[j + 1] = array[j];
        array[j] = temp;
        isSorted = false;
        lastExchangeIndex = j;
      }
    }
    sortedBorder = lastExchangeIndex;
    if(isSorted){
      break;
    }
  }
  return array;
}

File: myAlgorithm/001冒泡排序.java
//原始版本
private static int[] bubbleSort(int[] array){
  if(array == null || array.length <= 1){
    return array;
  }
  for(int i = 0; i < array.length - 1; i++){
    for(int j = 0; j < array.length - i - 1; j++){
      if(array[j] > array[j + 1]){
        int temp = array[j];
        array[j] = array[j + 1];
        array[j + 1] = temp;
      }
    }
  }
  return array;
}

//改进版本
private static int[] bubbleSort(int[] array){
  if(array == null || array.length <= 1){
    return array;
  }
  for(int i = 0; i < array.length - 1; i++){
    boolean isSorted = true;
    for(int j = 0; j < array.length - i - 1; j++){
      if(array[j] > array[j + 1]){
        int temp = array[j];
        array[j] = array[j + 1];
        array[j + 1] = temp;
        isSorted = false;
      }
    }
    if(isSorted){
      break;
    }
  }
  return array;
}

//终极改进版本
private static int[] bubbleSort(int[] array){
  if(array == null || array.length <= 1){
    return array;
  }
  int lastExchangeIndex = 0;
  int sortBorder = array.length - 1;
  for(int i = 0; i < array.length - 1; i++){
    boolean isSorted = true;
    for(int j = 0; j < sortBorder; j++){
      if(array[j] > array[j + 1]){
        int temp = array[j];
        array[j] = array[j + 1];
        array[j + 1] = temp;
        isSorted = false;
        lastExchangeIndex = j;
      }
    }
    sortBorder = lastExchangeIndex;
    if(isSorted){
      break;
    }
  }
  return array;
}
